### PR TITLE
Top level directory refactoring

### PR DESCRIPTION
**Pull request by gregschohn**
### Description

TODO: I will update jenkins settings on how to invoke gradle since gradlew has moved after/as this is merged.

Make gradle work at the top-level of the repo.  RFS and TrafficCapture settings have been unified into gradle settings at the root.  coreUtilities has also been migrated to the root so that it's easily visible and in a logical spot.  gradlew has also been moved.  I recommend installing the function below into .zshrc files so that 'agradle' can be invoked from any directory.

```
function agradle {                                     
  local dir=$(pwd)

  while [[ "$dir" != "/" ]]; do
    if [[ -x "$dir/gradlew" ]]; then
      "$dir/gradlew" "$@"
      return
    fi
    dir=$(dirname "$dir")
  done

  echo "No gradlew script found in the ancestor directories"
}
```

* Category Refactoring
* Why these changes are required? To make our projects more consistent and easier to develop.  More immediately, to easily pull in the coreUtilities instrumentation classes.
* What is the old behavior before changes and new behavior after changes?  RFS and TrafficCapture were disjoint gradle projects with separate settings.  Now one gradle project can understand both of them at the same time.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1738

### Testing
Manual tests w/ C&R w/ OSB.  All unit tests pass.

### Check List
- [ ] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    include projectName
+
+    project(projectName).projectDir = dir
+
+    for (File subdir : dir.listFiles()) {
+        addSubProjects(projectName, subdir)
+    }
+}
+
+rootProject.name = 'MigrationsAssistant'
```
Comment (peternied) [inline settings.gradle:40]: Seems like this should be called 'OpenSearch Migration' unless we are thinking of changing the repo name?

File: settings.gradle [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+include 'TrafficCapture:captureKafkaOffloader'
```
Comment (peternied) [inline settings.gradle:41]: Can we programmatically include these projects rather than list them?

File: settings.gradle [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+include 'TrafficCapture:captureOffloader'
+include 'TrafficCapture:captureProtobufs'
+include 'TrafficCapture:nettyWireLogging'
+include 'TrafficCapture:trafficCaptureDependencies'
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply settings.gradle:45]: They should and soon, but I wanted to leave the people working on RFS currently to make that call.  
I'm hoping that this commit/merge can open the door to changes like that but also not require them to be made immediately.

File: settings.gradle (hunk 2 above)
Comment (gregschohn) [reply settings.gradle:41]: How are you thinking that they could be programmatically included?

Comment (peternied) [reply settings.gradle:41]: Let me rephrase - do we have to enumerate each project in this way? Seems like we have a method to find all projects with `addSubProjects()` that could be used.

File: settings.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (peternied) [reply settings.gradle:45]: What do you think about moving the dependencies project to the top level now to minimize future changes on the projects under TrafficCapture.  Then how/when RFS moves it does on its own timeline?

[the remark this answers is not in the document]
Comment (gregschohn) [reply settings.gradle:45]: I'll move it up, but I'll need a new name other than trafficCaptureDependencies.  Does "commonDependencyVersionConstraints" work?  commonDependencies sounds like the project should be logging, coreUtilities, etc.  But I don't like the idea that we need any dependencies at all in any of our projects, hence the reason that I've added the extra words.

QQ: If these tools were in separate repos, would you still want to see common dependencies locked across repos/projects?  Would you have a replayer project, capture project, & RFS project all take a dependency on a common dependency project?  I think that it would still make sense, but it does (literally & intentionally) bind constraints on separate projects.

**Comment by chelma** [comment]
@gregschohn  LGTM, but - what does the gradle command look like to run the various components?  Is there namespacing or something?  Sorry for the noob question...

File: .github/workflows/e2eTest.yml [hunk 4]
```diff
@@ -30,7 +30,7 @@ jobs:
           cache-read-only: false
 
       - name: Start Docker Solution
-        run: ./gradlew dockerSolution:ComposeUp -x test --scan --info --stacktrace
+        run: ../gradlew dockerSolution:ComposeUp -x test --scan --info --stacktrace
         working-directory: TrafficCapture
```
Comment (peternied) [inline .github/workflows/e2eTest.yml:34]: I'd rather delete this line than use `..` above

Comment (peternied) [reply .github/workflows/e2eTest.yml:34]: duplicated in a couple of other places as well

File: RFS/README.md [hunk 5]
```diff
@@ -74,18 +74,18 @@ gradle run --args='-n global_state_snapshot --source-host $SOURCE_HOST --source-
 ### Using Docker
 RFS has support for packaging its java application as a Docker image by using the Dockerfile located in the `RFS/docker` directory. This support is directly built into Gradle as well so that a user can perform the below action, and generate a fresh Docker image (`migrations/reindex_from_snapshot:latest`) with the latest local code changes available.
 ```shell
-./gradlew buildDockerImages
+../gradlew buildDockerImages
```
Comment (peternied) [inline RFS/README.md:77]: What about indicating moving to the project root, or not including the path directive at all?
```suggestion
cd {PROJECT_ROOT}
./gradlew buildDockerImages
```

File: RFS/build-preloaded-source-image.gradle [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    dependsOn createInitialElasticsearchContainer
+    targetContainerId createInitialElasticsearchContainer.getContainerId()
+}
+
+task createClientContainer(type: DockerCreateContainer) {
+    dependsOn startInitialElasticsearchContainer, buildDockerImage_trafficGenerator
+    targetImageId 'migrations/osb_traffic_generator:latest'
+    containerName = "traffic-generator-container-${uniqueId}"
+    hostConfig.network = myNetworkName
+    //cmd = ['skip_dataset']
```
Comment (peternied) [inline RFS/build-preloaded-source-image.gradle:45]: Remove?

File: RFS/build-preloaded-source-image.gradle [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+// Task to commit the source container to create a new image that will include the loaded data
+def sourceContainerCommitTask = task commitSourceContainer() {
+    dependsOn waitClientContainer
+
+    doLast {
+        def client = createDockerClient()
+        def containerId = createInitialElasticsearchContainer.getContainerId().get()
+
+        println "Committing container ${containerId} to image my-new-primary-image:latest"
```
Comment (peternied) [inline RFS/build-preloaded-source-image.gradle:67]: parameterize this so it doesn't get out of date?  Note might need slightly more refactoring
```suggestion
        println "Committing container ${containerId} to image ${image}:${version}"
```

File: TrafficCapture/README.md [hunk 8]
```diff
@@ -60,8 +60,8 @@ Partial example output of the OpenSearch Benchmark tool:
 Running opensearch-benchmark w/ 'geonames' workload...
 
    ____                  _____                      __       ____                  __                         __
-/ __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
-/ / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
+  / __ \____  ___  ____ / ___/___  ____ ___________/ /_     / __ )___  ____  _____/ /_  ____ ___  ____ ______/ /__
+ / / / / __ \/ _ \/ __ \\__ \/ _ \/ __ `/ ___/ ___/ __ \   / __  / _ \/ __ \/ ___/ __ \/ __ `__ \/ __ `/ ___/ //_/
```
Comment (peternied) [inline TrafficCapture/README.md:64]: Revert these lines?

Comment (gregschohn) [reply TrafficCapture/README.md:64]: This corrects the output from OSB. It uses an italicized font and requires those spaces to make the big characters legible

File: .github/workflows/e2eTest.yml (hunk 4 above)
Comment (chelma) [reply .github/workflows/e2eTest.yml:34]: ...now I'm curious what the difference is!

Comment (chelma) [reply .github/workflows/e2eTest.yml:34]: Looked into it, not sure there is a practical difference here

File: RFS/README.md (hunk 5 above)
Comment (gregschohn) [reply RFS/README.md:77]: what if I added this to each of the three READMEs or the top README?  It's long, but it works all through the tree.

`agradle() { local dir=$(pwd); while [[ "$dir" != "/" ]]; do [[ -x "$dir/gradlew" ]] && "$dir/gradlew" "$@" && return; dir=$(dirname "$dir"); done; echo "No gradlew script found in the ancestor directories"; }`

File: RFS/build-preloaded-source-image.gradle (hunk 7 above)
Comment (gregschohn) [reply RFS/build-preloaded-source-image.gradle:67]: removed the print altogether

File: RFS/README.md (hunk 5 above)
Comment (peternied) [reply RFS/README.md:77]: This seems like it creates more complexity than ease of use.  IMO its much simpler to force everything to be done from the root of the project, at least when you can get away with a single purpose oriented repo

File: settings.gradle (hunk 2 above)
Comment (gregschohn) [reply settings.gradle:44]: This might be a good idea, but I have enough changes in this PR.  We can revisit this behavior change later.

File: RFS/README.md (hunk 5 above)
Comment (gregschohn) [reply RFS/README.md:77]: I agree that the agradle thing is scary and off-putting.  I appreciate that '..' is unambiguous and won't affect the user's current environment as `cd` would.

File: settings.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (peternied) [reply settings.gradle]: What do you think about a shorter `dependencyConstraints`? IMO when/if we get to another repo that might be a good time to reconsider the naming convention. 

I'm not too concerned when by looking at the project gradle its clear there is no actual source code and its consistently a single thing - a big list of dependency versions

**Comment by gregschohn** [comment]
In response to @chelma [here](https://github.com/opensearch-project/opensearch-migrations/pull/656#issuecomment-2120570104)... if you’re in the same directory as before, nothing changes other than the path/command to run gradlew - for which I use agradle.  However, now you can also build from anywhere in the tree by suffixing with the fully-qualified path.  i.e. :RFS:build from w/in TrafficCapture.

**Review by chelma — APPROVED** [review]
Thanks - LGTM!

File: .github/workflows/e2eTest.yml (hunk 4 above)
Comment (peternied) [reply .github/workflows/e2eTest.yml:34]: Note, I'm happy to revisit this post merge myself if I've got other ideas

File: RFS/README.md (hunk 5 above)
Comment (peternied) [reply RFS/README.md:77]: Resolving this thread, I'm happy to revisit this post merge myself if I've got other ideas

**Review by peternied — APPROVED** [review]
Looks good to merge, but it seems like we've got a failure with the link checker finding.  Can you exclude/fix this issue before merging?

```
### Errors in deployment/cdk/opensearch-service-migration/package-lock.json

* [ERR] [https://paypal.me/naturalintelligence](https://paypal.me/naturalintelligence) |
    Failed: Network error: error sending request for url (https://www.paypal.me/naturalintelligence)
```
